hindsight/io_modules/derived_stream.lua

```diff
@@ -0,0 +1,197 @@
+                    return -3, err
+                else
+                    return 0
+                end
+            end
+            return -4
+        end
+
+        local function timer_event(ns, shutdown)
+            if shutdown or last_insert + ticker_interval <= ns / 1e9 then
+                local err = insert_files()
+                if not err then
+                    batch_checkpoint_update()
+                end
+            end
+        end
+
+        return process_message, timer_event
+    elseif format == "protobuf" or format == "tsv" then
+        local output_path = read_config("output_path") or error("output_path must be set")
+        local files_cnt = 0
+
+        local process_message
+        local function timer_event(ns, shutdown)
+            -- no op
+        end
+
+        local function prune_open_files()
+            if files_cnt >= 10 then
+                local min = math.huge
+                local min_key
+                for k,v in pairs(files) do
+                    if v.time_t < min then
+                        min = v.time_t
+                        min_key = k
+                    end
+                end
+                files[min_key].fh:close()
+                files[min_key] = nil
+                files_cnt = files_cnt - 1
+            end
+        end
+
+        local function get_output_fh(ts, ext)
+            local day = math.floor(ts / (SEC_IN_DAY * 1e9))
+            local f = files[day]
+            if not f then
+                prune_open_files()
+                local date = os.date("%Y%m%d", ts / 1e9)
+                local filename = string.format("%s/%s_%s.%s", output_path, cfg_name, date, ext)
+                f = {fh = nil, time_t = 0}
+                local err
+                f.fh = assert(io.open(filename, "a"))
+                f.fh:setvbuf("no")
+                files[day] = f
+                files_cnt = files_cnt + 1
+            end
+            f.time_t = os.time()
+            return f.fh
+        end
+
+        if format == "protobuf" then
+            local hpb = require "derived_stream.heka_protobuf"
+            process_message = function ()
+                local fh = get_output_fh(read_message(ts_field), "log")
+                local msg = {Type = name, Fields = {}}
+                local ok, err = pcall(hpb.write_message, fh, msg, schema)
+                if not ok then
+                    return -1, err
+                end
+                return 0
+            end
+        else
+            local tsv = require "derived_stream.tsv"
+            local nil_value = read_config("nil_value") or ""
+            process_message =  function ()
+                local fh = get_output_fh(read_message(ts_field), "tsv")
+                tsv.write_message(fh, schema, nil_value)
+                return 0
+            end
+        end
+
+        return process_message, timer_event
+    else
+        error("invalid derived stream format: " .. format)
```

Comment:
I haven't done a review of this code, but while testing it I discovered that omitting the `format` setting will mean the format var is nil and this concatenation will fail. Should either use `tostring(format)` in the concatenation, or explicitly check for format being nil.